File: linda-logic/src/main/java/it/unibo/coordination/linda/logic/LogicTupleImpl.java
package it.unibo.coordination.linda.logic;

import alice.tuprolog.Struct;
import alice.tuprolog.Term;

import java.util.Objects;

final class LogicTupleImpl implements LogicTuple {

    private final Struct term;

    LogicTupleImpl(final Term term) {
        Objects.requireNonNull(term);
        if (term instanceof Struct && LogicTuple.getPattern().match(term)) {
            this.term = (Struct) term;
        } else {
            this.term = LogicTuple.getPattern(term);
        }
    }

    @Override
    public String toString() {
        return asTerm().toString();
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof LogicTuple
                && LogicTuple.equals(this, (LogicTuple) o);
    }

    @Override
    public int hashCode() {
        return LogicTuple.hashCode(this);
    }

    public Struct asTerm() {
        return term;
    }

    public Term getValue() {
        return asTerm().getArg(0);
    }
}

File: linda-logic/src/main/kotlin/it/unibo/coordination/linda/logic/LogicTupleImpl.kt
package it.unibo.coordination.linda.logic

import it.unibo.tuprolog.core.Struct
import it.unibo.tuprolog.core.Term
import it.unibo.tuprolog.unify.Unificator.Companion.matches


internal class LogicTupleImpl(term: Term) : LogicTuple {

    private val term: Struct by lazy {
        if (term is Struct && LogicTuple.pattern.matches(term)) {
            term
        } else {
            LogicTuple.getPattern(term)
        }
    }

    override val value: Term
        get() = asTerm()[0]

    override fun toString(): String = asTerm().toString()

    override fun equals(other: Any?): Boolean = other is LogicTuple && LogicTuple.equals(this, other as LogicTuple?)

    override fun hashCode(): Int = LogicTuple.hashCode(this)

    override fun asTerm(): Struct = term
}